data-plane/python-bindings/srpc/src/srpc/server.py

```diff
@@ -83,64 +91,101 @@ async def run(self):
+            logger.error(
+                f"{instance} no handler found for session {session_info.id} with destination {session_info.destination_name}",
+            )
+            return
+
+        async def generator():
+            try:
+                while True:
+                    session_recv, request_bytes = await local_app.receive(
+                        session=session_info.id,
+                    )
+
+                    if not rpc_handler.request_streaming:
+                        break
+
+                    if (
+                        session_recv.metadata.get("code") == str(code_pb2.OK)
+                        and not request_bytes
+                    ):
+                        logger.info("End of stream received")
+                        break
+
+                    request = rpc_handler.request_deserializer(request_bytes)
+                    context = Context.from_sessioninfo(session_recv)
+
+                    yield request, context
+            except Exception as e:
+                logger.error(f"Error receiving messages: {e}")
+                raise
+
+        if not rpc_handler.request_streaming:
+            # Read the request from slim
+            session_recv, request_bytes = await local_app.receive(
+                session=session_info.id,
             )
 
-            # build the context for the RPC call
-            context = Context.from_sessioninfo(session)
+            request = rpc_handler.request_deserializer(request_bytes)
+            context = Context.from_sessioninfo(session_recv)
+        else:
+            request, context = generator(), Context.from_sessioninfo(session_info)
 
-            # Call the RPC handler
-            if session_id.destination_name in self.handlers:
-                rpc_handler: Rpc = self.handlers[session_id.destination_name]
-                request = rpc_handler.request_deserializer(msg)
+        logger.info(f"Handling request {request} with context {context}")
+        if not rpc_handler.response_streaming:
+            logger.info("handling unary response")
 
-                logging.info(
-                    f"{instance} calling handler {rpc_handler.name} for session {session_id.id}",
-                )
+            # Call the handler with the request and context
+            response = await rpc_handler.handler(request, context)
+
+            logger.info(f"Response------->: {response}")
 
-                # Send the response back to the client
-                if rpc_handler.response_streaming:
-                    async for response in rpc_handler.handler(request, context):
-                        response_bytes = rpc_handler.response_serializer(response)
-                        await local_app.publish_to(session_info, response_bytes)
+            # Create generator to send the response
+            async def generator():
+                yield response
 
-                    return
+            response_generator = generator()
```

Comment:
Nice 👍 I wonder if we can make this is a closure function on server to reduce the complexity of the handle_session function:

```
def create_request_stream_generator(session_id):
	async def generator():
		...
	return generator
```

then in here:

```
if rpc_handler.request_streaming:
	input, context = self.create_request_stream_generator(session_id), Context.from_sessioninfo(session_info)
else:
	session_recv, request_bytes = await local_app.receive(...)
	input, context = rpc_handler.request_deserializer(request_bytes), Context.from_sessioninfo(session_recv)
```